source/transactions/tests/README.rst

```diff
@@ -162,6 +168,13 @@ Each YAML file has the following keys:
 
       - ``errorLabelsOmit``: A list of error label strings that the
         error is expected not to have.
+        
+      - ``state``: A map of expected session states. The keys can be
```

Comment:
Can this be moved one level up (alongside `result`)? I realize it's being introduced to assert the state of a session _after_ an error occurs, but I don't think it needs to be limited to that purpose. I could see this being used in non-error tests down the line to generally assert the session's state after an operation.

If so, I might suggest `resultingTransactionState` as a field name, which would immediately clarify whether we're asserting the session's transaction state _after_ the operation is executed.

On a separate note, I'm curious whether all drivers can actually implement this. The [Transactions spec](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst) doesn't define any public API for exposing a Session's internal state. For PHP, we did end up adding an [`isInTransaction()`](https://jira.mongodb.org/browse/PHPC-1231) method, which will assist in implementing the [Convenient API for Transactions spec](https://github.com/mongodb/specifications/blob/master/source/transactions-convenient-api/transactions-convenient-api.rst); however, that doesn't report the exact state.

Perhaps this opens a wider discussion about whether or not spec tests should be limiting to using public API?

Reply:
Good point regarding state being independent of the result, I can move it once the API is decided on per the other comments.

Reply:
Just realized this field is documented as a map and can support multiple assertions at a time. Since each operation can only be associated with a single session, I don't see a need for that. Moreover, I would have thought there no need to specify the session again since we could infer it from the operation's `arguments.session` field.

That said, I think there's a better approach beneath our noses here. @ShaneHarvey recently began adding [Special Test Operations](https://github.com/mongodb/specifications/blob/master/source/transactions/tests/README.rst#special-test-operations) to the transaction spec tests. I think it'd be preferable to introduce a new special operation to allow assertion of a session state. Splitting this assertion into its own operation also avoids any potential misunderstanding about whether the state should be inspected before or after the driver operation is executed -- this means there'd be no need to add more instruction to the already lengthy list in [Use as Integration Tests](https://github.com/mongodb/specifications/blob/master/source/transactions/tests/README.rst#use-as-integration-tests). 

Consider:

```
- name: assertSessionTransactionState
  object: testRunner
  arguments:
    session: session0
    state: in_progress
```

This is fairly self-documenting, and the fact that it'd a top-level operation means it'd be much easier for drivers to detect this unrecognized operation type (vs. the possibility of silently ignoring a new field in the `operations` document).

I went without the map syntax, as I still don't see a need for asserting multiple states at once; however, a test could easily perform multiple assertions by sequencing two of these operations. An additional benefit is that we could also assert the session's state before any operation is run (probably not needed for this PR, but potentially useful). Thoughts?

Reply:
Agreed, I like the `assertSessionTransactionState` idea!

For inspecting a ClientSession's state, I think some drivers may just be required to add public API like `session.getState()`. Other drivers, like python and maybe Java, may be able to keep it private.